### DISCUSSION TITLE
I find confusing that    select '12345'[1:3] and select '12345'[0:3]  produce the same results

### DISCUSSION THREAD

**Discussion by littlegreenbean33** [duckdb/duckdb#6858] [Q&A]
This applies as well to  substring/substr  where it is documented.

However, I would have expected an out of bound index.

Myself coming form the Python side of things.....

What's the rationale ?

```
 select '12345'[1:3]
> ;
┌──────────────┐
│ '12345'[1:3] │
│   varchar    │
├──────────────┤
│ 123          │
└──────────────┘
D select '12345'[0:3] ;
┌──────────────┐
│ '12345'[0:3] │
│   varchar    │
├──────────────┤
│ 123          │
└──────────────┘
```
Thanks.

JCG

**Comment by Tishj** [comment]
I think that's an attempt at backwards compatibility
Our indexes are 1-based, so [0:3] is not really valid
But it makes *some* sense that this works, because that's a start and a stop, so it likely just doesn't add anything for index 0.
[0:3:2] vs [1:3:2] should proof/disproof this theory

**Reply by littlegreenbean33** [reply]
thanks @Tishj  for taking the time to respond.

I think there is maybe an opportunity to be more explicit in the documentation, maybe including some examples where the behavior is different from the often expected Python results.

Cheers.

JCG

```
D select '12345'[3:-1] ;
┌───────────────┐
│ '12345'[3:-1] │
│    varchar    │
├───────────────┤
│ 34            │
└───────────────┘
D select '12345'[3:] ;
┌─────────────────┐
│ '12345'[3:NULL] │
│     varchar     │
├─────────────────┤
│ 345             │
└─────────────────┘

IPthon 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.11.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: "12345"[3:-1]
Out[1]: '4'

In [2]: "12345"[3:]
Out[2]: '45'


```


JCG

**Reply by Tishj** [reply]
The behavior is the same, only the indexing system is different, which is documented.
(<https://duckdb.org/docs/sql/functions/nested> this syntax is an alias for `list_slice`)
As you've shown:
```py
"12345"[3:-1]
"12345"[3:]
```
is equivalent to:
```sql
select '12345'[4:-1];
select '12345'[4:];
```

Also we sadly don't support providing a `step` yet (`[3:-1:2]` for example) so the test I came up with to proof/disproof my theory doesn't work.
But doing a select on [0] shows that this is likely the case, since it's empty
```sql
D select '12345'[0];
┌────────────┐
│ '12345'[0] │
│  varchar   │
├────────────┤
│            │
└────────────┘
```
